File: c++/0209.cpp
// Link: https://leetcode.com/problems/minimum-size-subarray-sum/description/
// Runtime: 38 ms
// Memory: 28.1 MB

class Solution 
{
public:
    int minSubArrayLen(int target, const std::vector<int>& nums) 
    {
        int sum = 0, left = 0, minimumSize = static_cast<int>(nums.size());
        bool hasCorrectSum = false;
        for (int right = 0; right < static_cast<int>(nums.size()); right++)
        {
            sum += nums[right];

            while (sum >= target)
            {
                minimumSize = std::min(minimumSize, right - left + 1);
                sum -= nums[left++];
                hasCorrectSum = true;
            }
        }

        if (hasCorrectSum) return minimumSize;
        else return 0;
    }
};

File: python/0209.py
# Link: https://leetcode.com/problems/minimum-size-subarray-sum/description/
# Runtime: 258 ms
# Memory: 29.1 MB


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        sum, left, minimum_size = 0, 0, len(nums)
        has_correct_sum = False
        for right, value in enumerate(nums):
            sum += value

            while sum >= target:
                minimum_size = min(minimum_size, right - left + 1)
                sum -= nums[left]
                left += 1
                has_correct_sum = True

        if has_correct_sum:
            return minimum_size
        else:
            return 0
